### PR TITLE
S3 integration

### PR DESCRIPTION
**Pull request by abhinadduri**

### CODE REVIEW

**Comment by dannycoates** [comment]
We should keep the file system version so we can dev without aws credentials. Lets factor out the storage bits into its own module. I sketched something up as a start here: https://gist.github.com/dannycoates/f75e37df25968ff74475156b894669d4

File: server/storage.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  module.exports = {
+    length: LocalLength,
+    get: LocalGet,
+    set: LocalSet,
+    delete: LocalDelete,
+    forceDelete: LocalForceDelete
+  };
+}
+
+function LocalLength(id) {
```
Comment (dannycoates) [inline server/storage.js]: Traditionally, in js this case style is used for constructors. Let's use camelCase for these function names

File: server/portal_server.js [hunk 2]
```diff
@@ -4,12 +4,24 @@ const path = require('path');
...
+  conf.bitly_key !== 'localhost';
+
+const AWS = require('aws-sdk');
+const s3 = new AWS.S3();
 
 const app = express();
-const redis = require('redis'),
-  client = redis.createClient();
+const redis = require('redis');
+const redis_client = redis.createClient();
```
Comment (dannycoates) [inline server/portal_server.js]: can we move all this redis code into `storage.js`?

File: server/portal_server.js [hunk 3]
```diff
@@ -4,12 +4,24 @@ const path = require('path');
...
+const fetch = require('node-fetch');
+const storage = require('./storage.js');
+
+let isProduction =
+  conf.env === 'production' &&
+  conf.s3_bucket !== 'localhost' &&
+  conf.bitly_key !== 'localhost';
+
+const AWS = require('aws-sdk');
+const s3 = new AWS.S3();
```
Comment (dannycoates) [inline server/portal_server.js]: We can remove `AWS` and `s3` from this file now

File: server/portal_server.js [hunk 4]
```diff
@@ -4,12 +4,24 @@ const path = require('path');
 const fs = require('fs-extra');
 const bodyParser = require('body-parser');
 const crypto = require('crypto');
+const conf = require('./config.js');
+const stream = require('stream');
+const fetch = require('node-fetch');
+const storage = require('./storage.js');
+
+let isProduction =
```
Comment (dannycoates) [inline server/portal_server.js]: It's confusing that `isProduction` can be `false` when `NODE_ENV=production`. Perhaps naming it something like `useRemoteServices` or `localhostOnly`.

File: server/storage.js [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+      });
+    });
+  });
+}
+
+function LocalDelete(id, client, delete_token) {
+  return new Promise((resolve, reject) => {
+    client.hget(id, 'delete', (err, reply) => {
+      if (!reply || delete_token !== reply) {
+        resolve(
```
Comment (dannycoates) [inline server/storage.js]: this block can just be `reject()`

File: server/storage.js [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  let params = {
+    Bucket: conf.s3_bucket,
+    Key: id,
+    Body: file
+  };
+
+  return new Promise((resolve, reject) => {
+    s3.upload(params, function(err, data) {
+      if (err) {
+        console.log(err, err.stack); // an error occurred
```
Comment (dannycoates) [inline server/storage.js]: We should call `reject()` here

File: server/storage.js [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    client.hget(id, 'delete', (err, reply) => {
+      if (!reply || delete_token !== reply) {
+        resolve(
+          new Promise((resolve, reject) => {
+            reject();
+          })
+        );
+      } else {
+        resolve(
+          new Promise((resolve, reject) => {
```
Comment (dannycoates) [inline server/storage.js]: we shouldn't need this `Promise`, just resolve the unlink

File: server/portal_server.js (hunk 4 above)
Comment (pdehaan) [inline server/portal_server.js]: Nit: I don't think this variable is used in this file anymore.
But @dannycoates comment still applies to server/storage.js:10 below.

**Comment by dannycoates** [comment]
This looks good to merge. I don't think we necessarily need index to be a template but NBD

**Pull request merged**